site/client/src/main/scala/Effects.scala

```diff
@@ -16,17 +18,29 @@ object Effects {
   def noop: Future[Option[Action]] = Future(None)
 
   def perform(s: State, a: Action): Future[Option[Action]] = a match {
-    case CompileExercise(method) ⇒ {
-      findByMethod(s, method) match {
-        case Some(exercise) if exercise.isFilled ⇒ Client.compileExercise(exercise).map(result ⇒ {
-          if (result.ok)
-            Some(CompilationOk(result.method))
-          else
-            Some(CompilationFail(result.method, result.msg))
-        })
-        case _ ⇒ noop
-      }
-    }
-    case _ ⇒ noop
+    case Start()                 ⇒ loadInitialData
+    case CompileExercise(method) ⇒ compileExercise(s, method)
+    case _                       ⇒ noop
+  }
+
+  def loadInitialData: Future[Option[Action]] = {
+    DomHandler.libraryAndSection.fold(Future(None): Future[Option[Action]])(libAndSection ⇒ {
+      val (lib, sect) = libAndSection
+      Client.fetchProgress(lib, sect).map(maybeState ⇒ {
```

Comment:
This may also be simplified via collect and partial function to avoid the doubly nested map